source/base/data_out_base.cc

```diff
@@ -7540,6 +7565,140 @@ namespace DataOutBase
+      &                                  nonscalar_data_ranges,
+    const Deal_II_IntermediateFlags &    flags,
+    const std::string &                  filename,
+    const MPI_Comm &                     comm,
+    const VtkFlags::ZlibCompressionLevel compression)
+  {
+#ifndef DEAL_II_WITH_MPI
+
+    (void)comm;
+#else
+
+    // We write a simple format based on the text format of
+    // write_deal_II_intermediate() on each MPI rank. The text format
+    // is quite verbose and we should probably change this to a more
+    // efficient binary representation at some point. The file layout
+    // is as follows:
+    //
+    // 1. A binary header with layout struct
+    //    ParallelIntermediateHeaderType.
+    // 2. A list of uint64_t with one value per rank denoting the
+    //    compressed size of the chunks of the next step.
+    // 3. The (potentially compressed) chunks as generated by
+    //    write_deal_II_intermediate() on each MPI rank.
+
+    // First generate my data by writing (optionally compressed) data into
+    // my_buffer:
+    std::vector<char> my_buffer;
+    {
+      boost::iostreams::filtering_ostream f;
+
+      if (compression != VtkFlags::no_compression)
+        f.push(boost::iostreams::zlib_compressor());
+
+      boost::iostreams::back_insert_device<std::vector<char>> inserter(
+        my_buffer);
+      f.push(inserter);
+
+      write_deal_II_intermediate<dim, spacedim>(
+        patches, data_names, nonscalar_data_ranges, flags, f);
+    }
+    const std::uint64_t my_size = my_buffer.size();
+
+    const unsigned int  my_rank     = Utilities::MPI::this_mpi_process(comm);
+    const std::uint64_t num_ranks   = Utilities::MPI::n_mpi_processes(comm);
+    const std::uint64_t num_patches = Utilities::MPI::sum(patches.size(), comm);
+
+    const ParallelIntermediateHeaderType header{
+      0x00dea111,
+      Deal_II_IntermediateFlags::format_version,
+      compression,
+      dim,
+      spacedim,
+      num_ranks,
+      num_patches};
+
+    // rank 0 also collects and writes the size of the data from each rank in
+    // bytes:
+    std::vector<std::uint64_t> chunk_sizes(num_ranks);
+    int                        ierr = MPI_Gather(
+      &my_size, 1, MPI_UINT64_T, chunk_sizes.data(), 1, MPI_UINT64_T, 0, comm);
+    AssertThrowMPI(ierr);
+
+    MPI_Info info;
+    MPI_Info_create(&info);
+    AssertThrowMPI(ierr);
+    MPI_File fh;
+    ierr = MPI_File_open(
+      comm, filename.c_str(), MPI_MODE_CREATE | MPI_MODE_WRONLY, info, &fh);
+    AssertThrow(ierr == MPI_SUCCESS, ExcFileNotOpen(filename));
+
+    ierr = MPI_File_set_size(fh, 0); // delete the file contents
+    AssertThrowMPI(ierr);
+    // this barrier is necessary, because otherwise others might already write
+    // while one core is still setting the size to zero.
+    ierr = MPI_Barrier(comm);
+    AssertThrowMPI(ierr);
+    ierr = MPI_Info_free(&info);
+    AssertThrowMPI(ierr);
```

Comment:
If I see this right, you only need the info object for the call to `MPI_File_open`, right? If so, move these two lines right below the call to `MPI_File_open`.